test/check-machines-consoles

```diff
@@ -744,6 +745,102 @@ fullscreen=0
         b.wait_visible(".vm-console-vnc canvas")
         self.assertIn("password=on", m.execute(f"grep ^-vnc /var/log/libvirt/qemu/{name}.log"))
 
+        # Change global password, but without restarting the VM. This
+        # means that Cockpit will use the new password when
+        # connecting, but the VM still expects the old.
+        m.execute("sed -i -e 's/vnc_password = \"barfoo\"/vnc_password = \"foobar\"/' /etc/libvirt/qemu.conf")
+
+        # We are still connected, but reconnecting will fail.
+
+        b.click(".consoles-card button:contains(Disconnect)")
+        b.wait_in_text(".consoles-card", "Disconnected")
+
+        # Quite often the reconnection here fails with code 1006
+        # (Abnormal Closure), but retrying will give us the expected
+        # error message, usually on the second try.
+
+        b.click(".consoles-card button:contains(Connect)")
+        for _ in range(5):
+            try:
+                b.wait_in_text(".consoles-card", "Authentication failed")
+                break
+            except testlib.Error:
+                if "Failed to connect" in b.text(".consoles-card"):
+                    print("Unexpected error message. Retrying...")
+                    b.click(".consoles-card button:contains(Retry)")
+                else:
+                    raise
+
+        b.wait_in_text(".consoles-card", "Authentication failed")
+
+    @testlib.skipImage("cockpit-ws too old",
+                       "rhel-9-*", "centos-9-*", "ubuntu-*", "debian-trixie", "opensuse-tumbleweed")
+    def testNonSharedExternal(self):
+        b = self.browser
+
+        name = "subVmTest1"
+        self.createVm(name, graphics="vnc")
+
+        self.login_and_go("/machines")
+        self.waitPageInit()
+        self.waitVmRow(name)
+        self.goToVmPage(name)
+
+        # VNC viewer should be connected
+        b.wait_visible(".vm-console-vnc canvas")
+
+        # Start external non-shared viewer.  We use Cockpit itself, in
+        # "Remote resizing" mode.
+        b2 = self.new_browser()
+        cookie = b.cookie("cockpit")
+        b2.open(f"/cockpit/@localhost/machines/index.html#/vm/vnc?name={name}&connection=system", cookie=cookie)
+        with b2.wait_timeout(60):
+            b2.select_PF("#vm-console-vnc-scaling", "Remote resizing")
+
+        # Our viewer should disconnect cleanly
+        b.wait_in_text(".consoles-card", "Disconnected")
+
+        # Trying to reconnect should fail while the external viewer is running
+        b.click('.consoles-card button:contains("Connect")')
+        b.wait_in_text(".consoles-card", "Failed to connect")
+
+        # Stopping the external viewer should allow us to connect again
+        b2.go("#/")
+        b.click('.consoles-card button:contains("Retry")')
```

Comment:
That looks like a race condition? I.e. `b2` needs to wait a bit until the disconnect actually happened. Is that visible with some `virsh` polling loop or similar? i.e. checking the server side for connected clients?

Reply:
Yes, good point.

Reply:
I added a "VNC connection counter" and check the expected number at various points.